File: selenium_java/pages/InventarioPagina.java
// POM - página de inventario
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import java.time.Duration;
import java.util.List;

public class InventarioPagina {
    private WebDriver driver;
    private WebDriverWait espera;
    private JavascriptExecutor js;

    public InventarioPagina(WebDriver driver) {
        this.driver = driver;
        this.espera = new WebDriverWait(driver, Duration.ofSeconds(5));
        this.js = (JavascriptExecutor) driver;
    }

    public void agregarProductos(int cantidad) {
        espera.until(ExpectedConditions.presenceOfAllElementsLocatedBy(
            By.cssSelector("[data-test^='add-to-cart']")));

        for (int i = 0; i < cantidad; i++) {
            List<WebElement> botones = driver.findElements(
                By.cssSelector("[data-test^='add-to-cart']"));

            if (botones.isEmpty()) break;

            WebElement boton = botones.get(0);

            js.executeScript(
                "arguments[0].scrollIntoView({behavior:'smooth', block:'center'});", boton);
            pausa(700);

            js.executeScript("arguments[0].style.outline='3px solid red';", boton);
            pausa(500);

            boton.click();
            if (i == 0) {
                espera.until(ExpectedConditions.visibilityOfElementLocated(
                    By.cssSelector(".shopping_cart_badge")));
            } else {
                espera.until(ExpectedConditions.textToBePresentInElementLocated(
                    By.cssSelector(".shopping_cart_badge"), String.valueOf(i + 1)));
            }
            pausa(800);
        }
    }

    public void agregarPrimerProductoAlCarrito() {
        agregarProductos(1);
    }

    public int obtenerContadorCarrito() {
        try {
            espera.until(ExpectedConditions.visibilityOfElementLocated(
                By.cssSelector(".shopping_cart_badge")));
            return Integer.parseInt(
                driver.findElement(By.cssSelector(".shopping_cart_badge")).getText());
        } catch (Exception e) {
            return 0;
        }
    }

    public void irAlCarrito() {
        WebElement icono = espera.until(ExpectedConditions.elementToBeClickable(
            By.cssSelector(".shopping_cart_link")));

        js.executeScript("arguments[0].style.outline='3px solid blue';", icono);
        pausa(600);

        icono.click();
        espera.until(ExpectedConditions.urlContains("cart"));
        pausa(1000);
    }

    private void pausa(long ms) {
        try { Thread.sleep(ms); } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
    }
}

File: selenium_python/paginas/InventarioPagina.py
# POM - página de inventario de SauceDemo
import time
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


class InventarioPagina:
    def __init__(self, driver):
        self.driver = driver
        self.espera = WebDriverWait(driver, 5)

    def agregar_productos(self, cantidad):
        self.espera.until(EC.presence_of_all_elements_located(
            (By.CSS_SELECTOR, "[data-test^='add-to-cart']")))

        for i in range(cantidad):
            botones = self.driver.find_elements(By.CSS_SELECTOR, "[data-test^='add-to-cart']")
            if not botones:
                break

            boton = botones[0]

            self.driver.execute_script(
                "arguments[0].scrollIntoView({behavior:'smooth', block:'center'});", boton)
            time.sleep(0.7)

            self.driver.execute_script("arguments[0].style.outline='3px solid red';", boton)
            time.sleep(0.5)

            boton.click()

            if i == 0:
                self.espera.until(EC.visibility_of_element_located(
                    (By.CSS_SELECTOR, ".shopping_cart_badge")))
            else:
                self.espera.until(EC.text_to_be_present_in_element(
                    (By.CSS_SELECTOR, ".shopping_cart_badge"), str(i + 1)))
            time.sleep(0.8)

    def agregar_primer_producto_al_carrito(self):
        self.agregar_productos(1)

    def obtener_contador_carrito(self):
        try:
            self.espera.until(EC.visibility_of_element_located(
                (By.CSS_SELECTOR, ".shopping_cart_badge")))
            return int(self.driver.find_element(By.CSS_SELECTOR, ".shopping_cart_badge").text)
        except Exception:
            return 0

    def ir_al_carrito(self):
        icono = self.espera.until(EC.element_to_be_clickable(
            (By.CSS_SELECTOR, ".shopping_cart_link")))
        self.driver.execute_script("arguments[0].style.outline='3px solid blue';", icono)
        time.sleep(0.6)
        icono.click()
        self.espera.until(EC.url_contains("cart"))
        time.sleep(1.0)
